include/targets/cuda/mma_tensor_op/gemm.cuh

```diff
@@ -2,25 +2,116 @@
 
 #include <algorithm>
 #include <array.h>
-
-#if (__COMPUTE_CAPABILITY__ == 700)
-
-#include <mma_tensor_op/hmma_m16n16k16_sm70.cuh>
-
-#else
-
-#include <mma_tensor_op/hmma_m16n8k8_sm80.cuh>
-
-#endif
+#include <mma_tensor_op/mma_dispatch.cuh>
+#include <pipeline.cuh>
 
 namespace quda
 {
   namespace mma
   {
     // return the size of the shared memory needed for MMA with block shape bM, bN, bK.
-    constexpr int shared_memory_bytes(int bM, int bN, int bK)
+    template <class mma_t> constexpr int shared_memory_bytes(int bM, int bN, int bK)
     {
-      return (bM + pad_size(bM) + bN + pad_size(bN)) * bK * 2 * sizeof(half);
+      return (bM + mma_t::pad_size(bM) + bN + mma_t::pad_size(bN)) * bK * 2 * sizeof(typename mma_t::compute_t);
+    }
+
+    /**
+      @brief Defining how many elements/atoms are there in type T ...
+     */
+    template <class T> struct batch_multiple {
+    };
+
+    /**
+      @brief ... e.g. there are 2 half's in a half2
+     */
+    template <> struct batch_multiple<half2> {
```

Comment:
We may as well `batch_multiple` out of this file and add it to `include/register_traits.h`

Reply:
https://github.com/lattice/quda/pull/1355/commits/ef5c9921a50a31bb86a0ed08739cb17ea6a85cca

Reply:
I have moved it back since batch_multiple needs `half2`, which is an CUDA specific thing...